pcsx2/Recording/InputRecording.cpp

```diff
@@ -465,364 +465,342 @@ wxString InputRecording::resolveGameName()
+// - persist last browsed IR path
+// - logs are weirdly formatted
+// - force OSD updates since a lot of input recording occurs during a paused state
+// - differentiating OSD logs somehow would be nice (color / a preceding icon?)
 
 void SaveStateBase::InputRecordingFreeze()
 {
 	// NOTE - BE CAREFUL
 	// CHANGING THIS WILL BREAK BACKWARDS COMPATIBILITY ON SAVESTATES
 	FreezeTag("InputRecording");
 	Freeze(g_FrameCount);
-
-	// Loading a save-state is an asynchronous task. If we are playing a recording
-	// that starts from a savestate (not power-on) and the starting (pcsx2 internal) frame
-	// marker has not been set (which comes from the save-state), we initialize it.
-	if (g_InputRecording.IsInitialLoad())
-		g_InputRecording.SetupInitialState(g_FrameCount);
-	else if (g_InputRecording.IsActive() && IsLoading())
-		g_InputRecording.SetFrameCounter(g_FrameCount);
 }
 
 InputRecording g_InputRecording;
 
-InputRecording::InputRecordingPad::InputRecordingPad()
+bool InputRecording::create(const std::string_view& fileName, const bool fromSaveState, const std::string_view& authorName)
 {
-	padData = new PadData;
+	if (!m_file.OpenNew(fileName, fromSaveState))
+	{
+		return false;
+	}
+
+	m_controls.setRecordMode();
+	if (fromSaveState)
+	{
+		std::string savestatePath = fmt::format("{}_SaveState.p2s", fileName);
+		if (FileSystem::FileExists(savestatePath.c_str()))
+		{
+			FileSystem::CopyFilePath(savestatePath.c_str(), fmt::format("{}.bak", savestatePath).c_str(), true);
+		}
+		m_initialSavestateLoadComplete = false;
+		m_type = Type::FROM_SAVESTATE;
+		m_isActive = true;
+		// TODO - error handling
+		VMManager::SaveState(savestatePath.c_str());
+	}
+	else
+	{
+		m_startingFrame = 0;
+		m_type = Type::POWER_ON;
+		m_isActive = true;
+		// TODO - should this be an explicit [full] boot instead of a reset?
+		VMManager::Reset();
+	}
+
+	m_file.getHeader().SetEmulatorVersion();
+	m_file.getHeader().SetAuthor(authorName);
+	m_file.getHeader().SetGameName(resolveGameName());
+	m_file.WriteHeader();
+	initializeState();
+	InputRec::log("Started new input recording");
+	InputRec::consoleLog(fmt::format("Filename {}", m_file.getFilename()));
+	return true;
 }
 
-InputRecording::InputRecordingPad::~InputRecordingPad()
+bool InputRecording::play(const std::string_view& filename)
 {
-	delete padData;
+	if (!m_file.OpenExisting(filename))
+	{
+		return false;
+	}
+
+	// Either load the savestate, or restart the game
+	if (m_file.FromSaveState())
+	{
+		std::string savestatePath = fmt::format("{}_SaveState.p2s", m_file.getFilename());
+		if (!FileSystem::FileExists(savestatePath.c_str()))
+		{
+			InputRec::consoleLog(fmt::format("Could not locate savestate file at location - {}", savestatePath));
+			InputRec::log("Savestate load failed");
+			m_file.Close();
+			return false;
+		}
+		m_type = Type::FROM_SAVESTATE;
+		m_initialSavestateLoadComplete = false;
+		m_isActive = true;
+		const auto loaded = VMManager::LoadState(savestatePath.c_str());
```

Comment:
Some concerns about threading here, if this function executes on the CPU thread, it's fine, but otherwise you'll need to use `Host::RunOnCPUThread().` (mainly because of the control calls below)